File: app/src/main/java/com/topjohnwu/magisk/view/dialogs/CustomAlertDialog.java
package com.topjohnwu.magisk.view.dialogs;

import android.content.Context;
import android.content.DialogInterface;
import android.view.LayoutInflater;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.annotation.StyleRes;
import androidx.appcompat.app.AlertDialog;

import com.topjohnwu.magisk.databinding.AlertDialogBinding;

public class CustomAlertDialog extends AlertDialog.Builder {

    private DialogInterface.OnClickListener positiveListener;
    private DialogInterface.OnClickListener negativeListener;
    private DialogInterface.OnClickListener neutralListener;

    protected AlertDialog dialog;
    protected AlertDialogBinding binding;

    {
        binding = AlertDialogBinding.inflate(LayoutInflater.from(getContext()));
        super.setView(binding.getRoot());
        binding.message.setVisibility(View.GONE);
        binding.negative.setVisibility(View.GONE);
        binding.positive.setVisibility(View.GONE);
        binding.neutral.setVisibility(View.GONE);
        binding.buttonPanel.setVisibility(View.GONE);
    }

    public CustomAlertDialog(@NonNull Context context) {
        super(context);
    }

    public CustomAlertDialog(@NonNull Context context, @StyleRes int themeResId) {
        super(context, themeResId);
    }

    @Override
    public CustomAlertDialog setView(int layoutResId) { return this; }

    @Override
    public CustomAlertDialog setView(View view) { return this; }

    @Override
    public CustomAlertDialog setMessage(@Nullable CharSequence message) {
        binding.message.setVisibility(View.VISIBLE);
        binding.message.setText(message);
        return this;
    }

    @Override
    public CustomAlertDialog setMessage(@StringRes int messageId) {
        return setMessage(getContext().getString(messageId));
    }

    @Override
    public CustomAlertDialog setPositiveButton(CharSequence text, DialogInterface.OnClickListener listener) {
        binding.buttonPanel.setVisibility(View.VISIBLE);
        binding.positive.setVisibility(View.VISIBLE);
        binding.positive.setText(text);
        positiveListener = listener;
        binding.positive.setOnClickListener(v -> {
            if (positiveListener != null) {
                positiveListener.onClick(dialog, DialogInterface.BUTTON_POSITIVE);
            }
            dialog.dismiss();
        });
        return this;
    }

    @Override
    public CustomAlertDialog setPositiveButton(@StringRes int textId, DialogInterface.OnClickListener listener) {
        return setPositiveButton(getContext().getString(textId), listener);
    }

    @Override
    public CustomAlertDialog setNegativeButton(CharSequence text, DialogInterface.OnClickListener listener) {
        binding.buttonPanel.setVisibility(View.VISIBLE);
        binding.negative.setVisibility(View.VISIBLE);
        binding.negative.setText(text);
        negativeListener = listener;
        binding.negative.setOnClickListener(v -> {
            if (negativeListener != null) {
                negativeListener.onClick(dialog, DialogInterface.BUTTON_NEGATIVE);
            }
            dialog.dismiss();
        });
        return this;
    }

    @Override
    public CustomAlertDialog setNegativeButton(@StringRes int textId, DialogInterface.OnClickListener listener) {
        return setNegativeButton(getContext().getString(textId), listener);
    }

    @Override
    public CustomAlertDialog setNeutralButton(CharSequence text, DialogInterface.OnClickListener listener) {
        binding.buttonPanel.setVisibility(View.VISIBLE);
        binding.neutral.setVisibility(View.VISIBLE);
        binding.neutral.setText(text);
        neutralListener = listener;
        binding.neutral.setOnClickListener(v -> {
            if (neutralListener != null) {
                neutralListener.onClick(dialog, DialogInterface.BUTTON_NEUTRAL);
            }
            dialog.dismiss();
        });
        return this;
    }

    @Override
    public CustomAlertDialog setNeutralButton(@StringRes int textId, DialogInterface.OnClickListener listener) {
        return setNeutralButton(getContext().getString(textId), listener);
    }

    @Override
    public AlertDialog create() {
        dialog = super.create();
        return dialog;
    }

    @Override
    public AlertDialog show() {
        create();
        dialog.show();
        return dialog;
    }

    public void dismiss() {
        dialog.dismiss();
    }
}

File: app/src/main/java/com/topjohnwu/magisk/view/dialogs/CustomAlertDialog.kt
package com.topjohnwu.magisk.view.dialogs

import android.content.Context
import android.content.DialogInterface
import android.view.LayoutInflater
import android.view.View
import androidx.annotation.StringRes
import androidx.annotation.StyleRes
import androidx.appcompat.app.AlertDialog

import com.topjohnwu.magisk.databinding.AlertDialogBinding

open class CustomAlertDialog : AlertDialog.Builder {

    private var positiveListener: DialogInterface.OnClickListener? = null
    private var negativeListener: DialogInterface.OnClickListener? = null
    private var neutralListener: DialogInterface.OnClickListener? = null

    protected var dialog: AlertDialog? = null
    protected var binding: AlertDialogBinding =
            AlertDialogBinding.inflate(LayoutInflater.from(context))

    init {
        super.setView(binding.root)
        binding.message.visibility = View.GONE
        binding.negative.visibility = View.GONE
        binding.positive.visibility = View.GONE
        binding.neutral.visibility = View.GONE
        binding.buttonPanel.visibility = View.GONE
    }

    constructor(context: Context) : super(context)

    constructor(context: Context, @StyleRes themeResId: Int) : super(context, themeResId)

    override fun setView(layoutResId: Int): CustomAlertDialog {
        return this
    }

    override fun setView(view: View): CustomAlertDialog {
        return this
    }

    override fun setMessage(message: CharSequence?): CustomAlertDialog {
        binding.message.visibility = View.VISIBLE
        binding.message.text = message
        return this
    }

    override fun setMessage(@StringRes messageId: Int): CustomAlertDialog {
        return setMessage(context.getString(messageId))
    }

    override fun setPositiveButton(text: CharSequence, listener: DialogInterface.OnClickListener?): CustomAlertDialog {
        binding.buttonPanel.visibility = View.VISIBLE
        binding.positive.visibility = View.VISIBLE
        binding.positive.text = text
        positiveListener = listener
        binding.positive.setOnClickListener {
            positiveListener?.onClick(dialog, DialogInterface.BUTTON_POSITIVE)
            dialog?.dismiss()
        }
        return this
    }

    override fun setPositiveButton(@StringRes textId: Int, listener: DialogInterface.OnClickListener?): CustomAlertDialog {
        return setPositiveButton(context.getString(textId), listener)
    }

    override fun setNegativeButton(text: CharSequence, listener: DialogInterface.OnClickListener?): CustomAlertDialog {
        binding.buttonPanel.visibility = View.VISIBLE
        binding.negative.visibility = View.VISIBLE
        binding.negative.text = text
        negativeListener = listener
        binding.negative.setOnClickListener {
            negativeListener?.onClick(dialog, DialogInterface.BUTTON_NEGATIVE)
            dialog?.dismiss()
        }
        return this
    }

    override fun setNegativeButton(@StringRes textId: Int, listener: DialogInterface.OnClickListener?): CustomAlertDialog {
        return setNegativeButton(context.getString(textId), listener)
    }

    override fun setNeutralButton(text: CharSequence, listener: DialogInterface.OnClickListener?): CustomAlertDialog {
        binding.buttonPanel.visibility = View.VISIBLE
        binding.neutral.visibility = View.VISIBLE
        binding.neutral.text = text
        neutralListener = listener
        binding.neutral.setOnClickListener {
            neutralListener?.onClick(dialog, DialogInterface.BUTTON_NEUTRAL)
            dialog?.dismiss()
        }
        return this
    }

    override fun setNeutralButton(@StringRes textId: Int, listener: DialogInterface.OnClickListener?): CustomAlertDialog {
        return setNeutralButton(context.getString(textId), listener)
    }

    override fun create(): AlertDialog {
        return super.create().apply { dialog = this }
    }

    override fun show(): AlertDialog {
        return create().apply { show() }
    }

    fun dismiss() {
        dialog?.dismiss()
    }
}
